Plutarch/Builtin.hs

```diff
@@ -39,6 +59,20 @@ ptailBuiltin = phoistAcyclic $ pforce $ punsafeBuiltin PLC.TailList
 pnullBuiltin :: Term s (PBuiltinList a :--> PBool)
 pnullBuiltin = phoistAcyclic $ pforce $ punsafeBuiltin PLC.NullList
 
+{- | A nil for builtin list of 'PData' elements.
+
+This uses 'PAsData' to preserve more information about the underlying 'PData'.
+-}
+pnilDataBuiltin :: Term s (PBuiltinList (PAsData a))
+pnilDataBuiltin = punsafeConstant . PLC.Some $ PLC.ValueOf (PLC.DefaultUniList PLC.DefaultUniData) []
+
+{- | A nil for builtin list of builtin pairs of 'PData' elements.
+
+This uses 'PAsData' to preserve more information about the underlying 'PData'.
+-}
+pnilPairDataBuiltin :: Term s (PBuiltinList (PBuiltinPair (PAsData a) (PAsData b)))
+pnilPairDataBuiltin = punsafeConstant . PLC.Some $ PLC.ValueOf (PLC.DefaultUniList $ PLC.DefaultUniPair PLC.DefaultUniData PLC.DefaultUniData) []
```

Comment:
While it's easy to add these functions, I don't think we should unfortunately, since we will have to deprecate them very soon. Notably, we want a generic `pconstant` that works for all combinations of literals as I've mentioned before. I can look into designing this if you're having a hard time imagining what it should look llike.

Reply:
Definitely want `pconstant` for all this stuff. Have that on my todo list a well. These here are really just meant to be alternatives that people can use in the meantime. As I have been myself in the validator rewrite.

Reply:
What is this function used for?

Reply:
I often find myself needing a `nil`. On the validator rewrites, I mostly use builtin list of `Data` - where I needed a `pnilDataBuiltin` to implement `filterData :: Term s (PBuiltinList a :--> PBuiltinList a)`.

I haven't needed `pnilPairDataBuiltin` it - but we'll need it when working with plutus assoc maps. The `Map` data value, `[(Data, Data)]`. This will end up representing things like `Value` - and implementing utilities on that `[(Data, Data)]` may require a `pnilPairDataBuiltin`.

Reply:
I am currently leaning toward exploring #53 a bit more before deciding on this.

Reply:
It'd be great to have #53 resolved so soon! No need to have specific `nil`s in that case. I'll remove these from this PR.

Reply:
With #62 you get write this functions as:

```haskell
pnil :: forall k a (s :: k). PLC.Contains PLC.DefaultUni (PDefaultUniType a) => Term s (PBuiltinList a)
pnil = pconstant @(PBuiltinList a) []

pnilPairData :: forall k (s :: k). (Data, Data) -> Term s (PBuiltinPair PData PData)
pnilPairData = pconstant @(PBuiltinPair PData PData)
```

Reply:
Why do you need the type applications?